File: java-backend/src/main/java/com/ajmservices/javabackend/model/Ressource.java
package com.ajmservices.javabackend.model;

import jakarta.persistence.*;

import java.util.Date;
import java.util.HashSet;
import java.util.Set;

import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import com.fasterxml.jackson.annotation.JsonIgnore;


@Entity
public class Ressource {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long idRessource;

    private String titre;
    private String description;
    private Date datePublication;
    private int nbConsultation;
    private int nbRecherche;
    private int nbPartage;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_utilisateur", nullable = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JsonIgnore
    private Utilisateur utilisateur;

    @ManyToOne
    @JoinColumn(name = "id_categorie")
    private Categorie categorie;

    @ManyToOne
    @JoinColumn(name = "id_type_ressource")
    private TypeRessource typeRessource;

    @ManyToOne
    @JoinColumn(name = "id_etat_ressource")
    private EtatRessource etatRessource;

    @OneToOne(mappedBy = "ressource", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Fichier fichier;

    @OneToMany(mappedBy = "ressource")
    private Set<Commentaire> commentaires = new HashSet<>();

    @ManyToMany
    @JoinTable(name = "_lier_ressource_type_relation", joinColumns = @JoinColumn(name = "idRessource"), inverseJoinColumns = @JoinColumn(name = "id_type_relation"))
    Set<TypeRelation> linkedTypeRelations;

    public Ressource(){

    };

    // public Ressource(String titre, String description, Date datePublication, int nbConsultation,
    //         int nbRecherche, int nbPartage, Utilisateur utilisateur, Categorie categorie, TypeRessource typeRessource,
    //         Fichier fichier) {

    public Ressource(String titre, String description, Date datePublication, int nbConsultation,
            int nbRecherche, int nbPartage, Utilisateur utilisateur, Categorie categorie, TypeRessource typeRessource,
            EtatRessource etatRessource) {

        this.titre = titre;
        this.description = description;
        this.datePublication = datePublication;
        this.nbConsultation = nbConsultation;
        this.nbRecherche = nbRecherche;
        this.nbPartage = nbPartage;
        this.utilisateur = utilisateur;
        this.categorie = categorie;
        this.typeRessource = typeRessource;
        this.etatRessource = etatRessource;
    };

    public Long getIdRessource() {
        return idRessource;
    }

    public void setIdRessource(Long idRessource) {
        this.idRessource = idRessource;
    }

    public String getTitre() {
        return titre;
    }

    public void setTitre(String titre) {
        this.titre = titre;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Date getDatePublication() {
        return datePublication;
    }

    public void setDatePublication(Date datePublication) {
        this.datePublication = datePublication;
    }

    public int getNbConsultation() {
        return nbConsultation;
    }

    public void setNbConsultation(int nbConsultation) {
        this.nbConsultation = nbConsultation;
    }

    public int getNbRecherche() {
        return nbRecherche;
    }

    public void setNbRecherche(int nbRecherche) {
        this.nbRecherche = nbRecherche;
    }

    public int getNbPartage() {
        return nbPartage;
    }

    public void setNbPartage(int nbPartage) {
        this.nbPartage = nbPartage;
    }

    public Utilisateur getUtilisateur() {
        return utilisateur;
    }

    public void setUtilisateur(Utilisateur utilisateur) {
        this.utilisateur = utilisateur;
    }

    public Fichier getFichier() {
        return fichier;
    }

    // Setter pour fichier
    public void setFichier(Fichier fichier) {
        this.fichier = fichier;
    }

    public Set<Commentaire> getCommentaires() {
        return commentaires;
    }

    // toString method
    @Override
    public String toString() {
        return "Ressource{" +
                "idRessource=" + idRessource +
                ", titre='" + titre + '\'' +
                ", description='" + description + '\'' +
                ", datePublication='" + datePublication + '\'' +
                ", nbConsultation='" + nbConsultation + '\'' +
                ", nbPartage='" + nbPartage + '\'' +
                ", nbRecherche=" + nbRecherche + '\'' +
                ", utilisateur=" + utilisateur + '\'' +
                ", categorie=" + categorie + '\'' +
                ", typeRessource=" + typeRessource + '\'' +
                ", fichier=" + fichier + '\'' +
                ", commentaires=" + commentaires +
                '}';
    }
}

File: symfony-backend/backend-ress-rela/src/Entity/Ressource.php
<?php

namespace App\Entity;

use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use App\Repository\RessourceRepository;
use Symfony\Component\Serializer\Annotation\Groups;

#[ORM\Entity(repositoryClass: RessourceRepository::class)]
class Ressource
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    #[Groups(["Utilisateur","Ressource", "Categorie", "TypeRelation", "EtatRessource", "TypeRessource"])]
    private ?int $id = null;

    #[Groups(["Utilisateur","Ressource", "Categorie", "TypeRelation", "EtatRessource", "TypeRessource"])]
    #[ORM\Column(length: 500)]
    private ?string $titre = null;

    #[Groups(["Utilisateur","Ressource", "Categorie", "TypeRelation", "EtatRessource", "TypeRessource"])]
    #[ORM\Column(length: 9999)]
    private ?string $description = null;

    #[Groups(["Utilisateur","Ressource", "Categorie", "TypeRelation", "EtatRessource", "TypeRessource"])]
    #[ORM\Column(type: Types::DATE_MUTABLE)]
    private ?\DateTimeInterface $datePublication = null;

    #[ORM\Column]
    private ?int $nbConsultation = 0;

    #[ORM\Column]
    private ?int $nbRecherche = 0;

    #[ORM\Column]
    private ?int $nbPartage = 0;

    #[ORM\ManyToOne(inversedBy: 'ressources')]
    #[ORM\JoinColumn(nullable: false)]
    #[Groups(["Ressource", "Categorie", "TypeRelation", "EtatRessource", "TypeRessource"])]
    private ?Utilisateur $utilisateur = null;

    #[ORM\ManyToOne(inversedBy: 'ressources')]
    #[ORM\JoinColumn(nullable: false)]
    #[Groups(["Ressource"])]
    private ?Categorie $categorie = null;

    /**
     * @var Collection<int, TypeRelation>
     */
    #[ORM\JoinTable(name:"_ressource_type_relation")]
    #[ORM\ManyToMany(targetEntity: TypeRelation::class, inversedBy: 'ressources')]
    #[Groups(["Ressource"])]
    private Collection $typeRelations;

    #[ORM\ManyToOne(inversedBy: 'ressources')]
    #[ORM\JoinColumn(nullable: false)]
    private ?EtatRessource $etatRessource = null;

    #[ORM\ManyToOne(inversedBy: 'ressources')]
    #[ORM\JoinColumn(nullable: false)]
    private ?TypeRessource $typeRessource = null;


    public function __construct()
    {
        $this->typeRelations = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getTitre(): ?string
    {
        return $this->titre;
    }

    public function setTitre(string $titre): static
    {
        $this->titre = $titre;

        return $this;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(string $description): static
    {
        $this->description = $description;

        return $this;
    }

    public function getDatePublication(): ?\DateTimeInterface
    {
        return $this->datePublication;
    }

    public function setDatePublication(\DateTimeInterface $datePublication): static
    {
        $this->datePublication = $datePublication;

        return $this;
    }

    public function getNbConsultation(): ?int
    {
        return $this->nbConsultation;
    }

    public function setNbConsultation(int $nbConsultation): static
    {
        $this->nbConsultation = $nbConsultation;

        return $this;
    }

    public function getNbRecherche(): ?int
    {
        return $this->nbRecherche;
    }

    public function setNbRecherche(int $nbRecherche): static
    {
        $this->nbRecherche = $nbRecherche;

        return $this;
    }

    public function getNbPartage(): ?int
    {
        return $this->nbPartage;
    }

    public function setNbPartage(int $nbPartage): static
    {
        $this->nbPartage = $nbPartage;

        return $this;
    }

    public function getUtilisateur(): ?Utilisateur
    {
        return $this->utilisateur;
    }

    public function setUtilisateur(?Utilisateur $utilisateur): static
    {
        $this->utilisateur = $utilisateur;

        return $this;
    }

    public function getCategorie(): ?Categorie
    {
        return $this->categorie;
    }

    public function setCategorie(?Categorie $categorie): static
    {
        $this->categorie = $categorie;

        return $this;
    }

    /**
     * @return Collection<int, TypeRelation>
     */
    public function getTypeRelations(): Collection
    {
        return $this->typeRelations;
    }

    public function addTypeRelation(TypeRelation $typeRelation): static
    {
        if (!$this->typeRelations->contains($typeRelation)) {
            $this->typeRelations->add($typeRelation);
        }

        return $this;
    }

    public function removeTypeRelation(TypeRelation $typeRelation): static
    {
        $this->typeRelations->removeElement($typeRelation);

        return $this;
    }

    public function getEtatRessource(): ?EtatRessource
    {
        return $this->etatRessource;
    }

    public function setEtatRessource(?EtatRessource $etatRessource): static
    {
        $this->etatRessource = $etatRessource;

        return $this;
    }

    public function getTypeRessource(): ?TypeRessource
    {
        return $this->typeRessource;
    }

    public function setTypeRessource(?TypeRessource $typeRessource): static
    {
        $this->typeRessource = $typeRessource;

        return $this;
    }

}
